Add infrastructure for managing system indices

pugnascotia opened the pull request:
Part of #61656.

Add the necessary support for automatically creating and updating system
indices. This works by making it possible to create a system index
descriptor with all the information needed to manage the mappings,
settings and aliases.

Follow-up work will opt existing indices into this framework.

pugnascotia committed:
Add infrastructure for managing system indices

Part of #61656.

Add the necessary support for automatically creating and updating system
indices. This works by making it possible to create a system index
descriptor with all the information needed to manage the mappings,
settings and aliases.

Follow-up work will opt existing indices into this framework.

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

pugnascotia committed:
Checkstyle

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -141,12 +152,56 @@ public ClusterState execute(ClusterState currentState) throws Exception {
+
+                    final SystemIndexDescriptor descriptor = systemIndices.findMatchingDescriptor(indexName);
+
+                    if (descriptor != null && descriptor.isAutomaticallyManaged()) {
```
Can we simplify this method a bit? We don't really share logic between non-system index creation and system index creation as far as I can see so I think we can test for the non system case and simply return the update request there?

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java:
```diff
@@ -63,13 +73,48 @@ protected void masterOperation(Task task, final CreateIndexRequest request, fina
+                .aliases(aliases)
                 .waitForActiveShards(request.waitForActiveShards());
 
+        if (isSystemIndex) {
```
Similar to the comment in `AutoCreateAction`, can we break up the system and non-system case since we don't really need to share anything between the two cases?

jaymode commented on server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:
```diff
@@ -0,0 +1,256 @@
+            return;
+        }
+
+        getEligibleDescriptors(state.getMetadata()).stream()
```
I think we need to make an effort to wait for any pending updates to finish prior to launching more by tracking them between cluster state updates. We've had issues in the past where we overloaded master nodes with too many requests due to a pattern like this.

jaymode reviewed (changes requested):
I took a brief look through the PR and left a few comments.

pugnascotia committed:
Address review feedback

- Separate logic for building create index requests into sytem-index and
  non-system-index versions
- Use an AtomicBoolean to prevent updates to system index mappings
  overloading the master

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-system-indices-infra

pugnascotia commented:
Note to reviewers: I'm holding off creating follow-up PRs to port over existing systems indices until this is merged, so I'd appreciate any and all feedback. In particular, some guidance would be good for setting up a BWC test that exercises the mappings upgrade logic in `SystemIndexManager`.

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -141,12 +152,51 @@ public ClusterState execute(ClusterState currentState) throws Exception {
+                        updateRequest.aliases(Set.of(new Alias(aliasName)));
+                    }
+
+                    logger.info("Auto-creating system index {}", concreteIndexName);
```
I think we should make this a debug log like the regular message

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java:
```diff
@@ -75,9 +87,24 @@ protected ClusterBlockException checkBlock(UpdateSettingsRequest request, Cluste
+                    .stream()
+                    .map(entry -> "[" + entry.getKey() + "] -> " + entry.getValue())
+                    .collect(Collectors.joining(", "));
+            logger.warn(message);
```
On backport, we will need to make this a deprecation warning

pugnascotia replied:
Can I clarify - in the backport, should this raise a deprecation warning but not fail the request?

jaymode replied:
Correct on backport this should result in a deprecation warning and still allow the request to go through.

jaymode commented on server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/TransportPutMappingAction.java:
```diff
@@ -87,12 +92,25 @@ protected void masterOperation(Task task, final PutMappingRequest request, final
+                    + violations
+                    + ": system indices can only use mappings from their descriptors, "
+                    + "but the mappings in the request did not match those in the descriptors(s)";
+                logger.warn(message);
```
On backport, we will need to make this a deprecation warning

jaymode commented on server/src/main/java/org/elasticsearch/common/Strings.java:
```diff
@@ -874,6 +874,13 @@ public static String cleanTruncate(String s, int length) {
         return s.substring(0, length);
     }
 
+    public static String requireNonEmpty(String str, String message) {
```
Can you add a javadocs for this method to indicate what it does and what is thrown?

jaymode commented on server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java:
```diff
@@ -81,10 +191,160 @@ public String getDescription() {
+        private int indexFormat = 0;
+        private String versionMetaKey = null;
+        private String origin = null;
+
```
How about adding a private constructor so that all creations are from `SystemIndexDescriptor.builder()`?

jaymode commented on server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java:
```diff
@@ -81,10 +191,160 @@ public String getDescription() {
+        public SystemIndexDescriptor build() {
+            String mappings = mappingsBuilder == null ? null : Strings.toString(mappingsBuilder);
+
+            Strings.requireNonEmpty(indexPattern, "indexPattern must be supplied");
```
Can all validation be moved into the constructor?

jaymode commented on server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:
```diff
@@ -0,0 +1,265 @@
+        if (state.nodes().isLocalNodeElectedMaster() == false) {
+            return;
+        }
+
```
Let's avoid extra work here by guarding this with a `if (upgradeInProgress.get() == false)` check

jaymode commented on server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:
```diff
@@ -0,0 +1,265 @@
+        originSettingClient.admin().indices().putMapping(request, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                isUpgradeInProgress.set(false);
```
this sets to false early. There can still be in flight items and isUpgradeInProgress would be incorrect. You may want to use something like a `GroupedActionListener`

pugnascotia replied:
I'm not sure I follow. The `upgradeIndexMetadata` method is called once for each index that needs upgraded, and the first thing the method does is attempt to set `isUpgradeInProgress` to `true`. So only one index will be upgraded, and the rest need to wait until `isUpgradeInProgress` becomes false again. I wrote it that way so as to be cautious about adding load to the master.

Were you thinking that we'd dispatch all the mapping updates in one go, and use a `GroupedActionListener` to release the `isUpgradeInProgress` once they'd all finished?

jaymode replied:
> I'm not sure I follow. 

That's because I did not follow the code properly. I think it would be best to fire off all of the upgrades at once rather than sequentially since IIRC the code is computing everything that needs upgraded each time.

> Were you thinking that we'd dispatch all the mapping updates in one go, and use a GroupedActionListener to release the isUpgradeInProgress once they'd all finished?

Yes that is what I was thinking. I wanted to avoid resending each one every time there was a cluster state update if other cluster state updates happened prior to the mapping updates that we had previously sent.

pugnascotia replied:
I've reworked the locking and upgrading to use a `GroupedActionListener`.

jaymode reviewed (changes requested):
I left some more feedback. I think this is close. Regarding testing overall, I think it might be hard without an index that uses this framework. So I think once we get this PR in and migrate one to use this then we can add tests that attempt to put mappings or settings that would trigger a violation.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-system-indices-infra

pugnascotia committed:
Review feedback

pugnascotia committed:
Apply all mappings updates at once

Run all upgrades at once, and use a GroupedActionListener to release the
lock once all have completed.

jaymode commented on server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:
```diff
@@ -83,9 +84,22 @@ public void clusterChanged(ClusterChangedEvent event) {
+                );
+
+                descriptors.forEach(descriptor -> upgradeIndexMetadata(descriptor, listener));
+            }
```
```suggestion
            } else {
                isUpgradeInProgress.set(false);
            }
```

pugnascotia replied:
Good catch! 😅

jaymode reviewed (approved):
Left one comment. Otherwise LGTM. Thanks for the iterations!

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 61656-auto-create-system-indices-infra

pugnascotia committed:
Remember to release the lock
